### DISCUSSION TITLE
how to use ssl certificate

### DISCUSSION THREAD

**Discussion by morningstarsabrina** [zauberzeug/nicegui#387] [Q&A] [answered]
Hello,

how to run nicegui with ssl (certificate,key) ?
I know uvicorn accept  'ssl_keyfile' and   'ssl_certfile' parameters, just like below
```py
uvicorn.run(app, host="0.0.0.0", port=8000,  ssl_keyfile="./key.pem",  ssl_certfile="./cert.pem")
```
is it possible to do the same thing with 'ui.run()' ?

thanks

**Comment by morningstarsabrina** [comment]
so I tried to run nicegui by using uvicorn.run() instead of ui.run() but I had a lot of errors. I think the problem is the app parameter. any idea how to do that ?

**Comment by falkoschindler** [comment, marked as answer]
You can adapt the [FastAPI example](https://github.com/zauberzeug/nicegui/tree/main/examples/fastapi) to start the server yourself:

```py
import uvicorn
from fastapi import FastAPI
from nicegui import ui

# you need to create the FastAPI app yourself
app = FastAPI()

# normal NiceGUI code
ui.label('Hello!')

# replace ui.run with ui.run_with to connect NiceGUI with the app
ui.run_with(app)

# start uvicorn with the app
uvicorn.run(app, host='0.0.0.0', port=8000)
```

Nevertheless, we should think about adding SSL parameters (or `**kwargs`) to `ui.run`.

**Reply by morningstarsabrina** [reply]
thanks that worked for me.
I agree about adding the SSL parameters to `ui.run()` because its very important feature

**Reply by morningstarsabrina** [reply]
Can adding SSL parameters to `ui.run` be tracked to be implemented later please ?

**Reply by falkoschindler** [reply]
Ok, done: #421.

**Reply by morningstarsabrina** [reply]
thank you very much
